File: Percolation/Percolation.js
const Grid = require('./Grid');

/* TODO: Add static helper*/

module.exports = class Percolation {
  //creates a grid
  constructor(n) {
    this.grid = new Grid(n);
    this.gridSize = n;
  }

  // opens the site (row, col) if it is not open already
  open(row, col) {
    let node = this._findNode(row, col);
    if (!node.open) {
      node.open = true;
      this.grid.openNodes++;
    }
  }

  //look at two nodes, both open? link them
  union(node1, node2) {
    if (node1.open && node2.open) {
      console.log(
        `about to open up ${(node1.row, node1.col)}, ${(node1.row, node2.col)} `
      );
      if (node1.root === node2.root) return;

      if (node1.root === 0 || node2.root === 0) {
        node1.root = 0;
        node2.root = 0;
      }

      if (node1.size < node2.size) {
        node1.root = node2.root;
        node2.size += node1.root;
      } else {
        node2.root = node1.root;
        node1.size += node2.root;
      }
    }
  }

  //created spaghetti code, refactoring
  checkFull(row, col) {
    let node = this._findNode(row, col);
    if (node.open === false) {
      return false;
    }

    if (node.full) {
      return true;
    }

    if (node.row === 0 && node.open) {
      node.full = true;
      return true;
    }
    let nodeUp = this._nodeAbove(node.row, node.col);

    let nodeLeft, nodeRight, nodeDown;
    // let answer;

    if (node.row > 0) {
      nodeLeft = this._nodeLeft(node.row, node.col);
    }
    if (node.row < this.gridSize) {
      nodeRight = this._nodeRight(node.row, node.col);
    }
    if (node.col < this.gridSize) {
      nodeDown = this._nodeBelow(node.row, node.col);
    }

    if (nodeUp) {
      console.log(
        `NODE ABOVE CHECK ${nodeUp.row}, ${nodeUp.col} - ${nodeUp.full}`
      );
      if (nodeUp.full) {
        node.full = true;
        this.answer = true;
      }
    }

    if (nodeLeft) {
      console.log(`NODE LEFT CHECK ${nodeLeft.row}, ${nodeLeft.col}`);
      if (nodeLeft.full) {
        node.full = true;
        this.answer = true;
      }
    }

    if (nodeRight) {
      console.log(`NODE RIGHT CHECK ${nodeRight.row}, ${nodeRight.col}`);
      if (nodeRight.full) {
        node.full = true;
        this.answer = true;
      }
    }

    if (nodeDown) {
      console.log(`NODE BOTTOM CHECK ${nodeDown.row}, ${nodeDown.col}`);
      if (nodeDown.full) {
        node.full = true;
        this.answer = true;
      }
    }
  }

  // returns the number of open sites

  numberOfOpenSites() {
    return this.grid.openNodes;
  }

  // does the system percolate?
  percolates() {
    //Does an individual node on the bottom row touch full nodes all the way to the top

    for (let i = 0; i < this.gridSize; i++) {
      let checkNode = this._findNode(this.gridSize - 1, i);
      // console.log(checkNode);
      if (this.checkFull(checkNode)) {
        console.log('IT PERCOLATES~~~~~~~~~~~~~~~~~~~~~~~~~~~');
        return true;
      }
    }

    return false;
  }

  _print() {
    this.grid ? this.grid.print() : null;
  }

  _findNode(row, col) {
    if (
      row >= 0 &&
      row <= this.gridSize - 1 &&
      col >= 0 &&
      col <= this.gridSize - 1
    ) {
      return this.grid.nodes[this._findNodeNumber(row, col)];
    }
  }

  _findNodeNumber(row, col) {
    return row * this.gridSize + col;
  }

  _nodeAbove(row, col) {
    return this._findNode(row - 1, col);
  }

  _nodeBelow(row, col) {
    return this._findNode(row + 1, col);
  }

  _nodeLeft(row, col) {
    return this._findNode(row, col - 1);
  }

  _nodeRight(row, col) {
    return this._findNode(row, col + 1);
  }
};

File: percolationJava/Percolation.java
/* *****************************************************************************
 *  Name:    George Gardner
 *  NetID:   Nope
 *  Precept: P00
 *
 *  Description:  Percolation class, for percolating
 *
 **************************************************************************** */

import edu.princeton.cs.algs4.In;
import edu.princeton.cs.algs4.WeightedQuickUnionUF;

public class Percolation {
  private final int top = 0;
  private final int bottom;
  private final boolean[][] opened;
  private final int size;
  private int openSites;
  private final WeightedQuickUnionUF uf;


  // creates n-by-n grid, with all sites initially blocked
  public Percolation(int n) {
    if (n <= 0) {
      throw new IllegalArgumentException();
    }
    size = n;
    bottom = size * size + 1;
    uf = new WeightedQuickUnionUF((size * size + 2));
    opened = new boolean[size][size];
    openSites = 0;

  }

  // opens the site (row, col) if it is not open already
  public void open(int row, int col) {
    checkException(row, col);
    int thisNode = getQuickFindIndex(row, col);

    opened[row][col] = true;
    ++openSites;

    // check top
    if (row == 0) {
      System.out.println("top check - " + thisNode + " " + top);
      uf.union(thisNode, top);
    }

    // check bottom
    if (row == size - 1) {
      System.out.println("bottom check - " + thisNode + " " + bottom);
      uf.union(thisNode, bottom);
    }

    // check above
    if (row > 0 && isOpen(row - 1, col)) {
      uf.union(thisNode, getQuickFindIndex(row - 1, col));
    }

    // check below
    if (row < size - 1 && isOpen(row + 1, col)) {
      uf.union(thisNode, getQuickFindIndex(row + 1, col));
    }

    // check left
    if (col > 0 && isOpen(row, col - 1)) {
      uf.union(thisNode, getQuickFindIndex(row, col - 1));
    }

    // check right
    if (col < size - 1 && isOpen(row, col + 1)) {
      uf.union(thisNode, getQuickFindIndex(row, col + 1));
    }


  }

  //  check to make sure row,col are inbounds
  private void checkException(int row, int col) {
    if (row < 0 || row > size || col < 0 || col > size) {
      throw new IllegalArgumentException();
    }
  }


  // is the site (row, col) open?
  public boolean isOpen(int row, int col) {
    if (opened[row][col]) {
      return true;
    }
    return false;
  }

  // is the site (row, col) full?
  public boolean isFull(int row, int col) {
    checkException(row, col);
    // System.out.println("Check full for: " + row + " " + col);
    return (uf.find(top) == uf.find(getQuickFindIndex(row, col)));

  }

  // returns the number of open sites
  public int numberOfOpenSites() {
    return openSites;
  }

  // does the system percolate?
  public boolean percolates() {
    // System.out.println(uf.find(top) + " " + uf.find(bottom) + " percolateS?");
    return (uf.find(top) == uf.find(bottom));

  }

  // find the index of the given row, col
  private int getQuickFindIndex(int row, int col) {
    return (size * row) + col + 1;
  }

  public void printGrid(int n) {
    for (int row = 0; row < n; row++) {
      for (int col = 0; col < n; col++) {
        System.out.printf("%d,%d\n ", row, col);
      }
    }
  }

  // test client (optional)
  public static void main(String[] args) {
    String filename = args[0];
    In in = new In(filename);
    int n = in.readInt();
    System.out.println(n);
    Percolation percolation = new Percolation(n);
    while (!in.isEmpty()) {
      int row = in.readInt();
      System.out.println(row);
      int col = in.readInt();
      System.out.println(col);
      percolation.open(row, col);
    }
    percolation.printGrid(n);
    System.out.println(percolation.percolates());
  }

}

